src/main/scala/is/hail/methods/Skat.scala

```diff
@@ -0,0 +1,318 @@
+package is.hail.methods
+
+import is.hail.utils._
+import is.hail.variant._
+import is.hail.expr._
+import is.hail.keytable.KeyTable
+import is.hail.stats.{RegressionUtils, SkatModel}
+import is.hail.annotations.Annotation
+import breeze.linalg._
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.Row
+
+object SkatStat {
+  val schema = TStruct(
+    ("q", TFloat64),
+    ("pval", TFloat64),
+    ("fault", TInt32))
+}
+
+case class SkatStat(q: Double, pval: Double, fault: Int) {
+  def toAnnotation = Annotation(q, pval, fault)
+}
+
+case class SkatTuple[T <: Vector[Double]](q: Double, xw: T, qtxw: DenseVector[Double])
+
+object Skat {
+
+  def apply(vds: VariantDataset,
+    keyName: String,
+    variantKeys: String,
+    singleKey: Boolean,
+    weightExpr: Option[String],
+    yExpr: String,
+    covExpr: Array[String],
+    useDosages: Boolean,
+    useLargeN: Boolean = false): KeyTable = {
+
+
+    def SkatRDDtoKeyTable[T <: Vector[Double]](keyedRdd:  RDD[(Any, Iterable[(T, Double)])], keyType: Type,
+      y: DenseVector[Double], cov: DenseMatrix[Double], keyName: String,
+      resultOp: (Array[SkatTuple[T]], Double) => SkatStat): KeyTable = {
+      val n = y.size
+      val k = cov.cols
+      val d = n - k
+
+      if (d < 1)
+        fatal(s"$n samples and $k ${ plural(k, "covariate") } including intercept implies $d degrees of freedom.")
+
+      // fit null model
+      val qr.QR(q, r) = qr.reduced.impl_reduced_DM_Double(cov)
+      val beta = r \ (q.t * y)
+      val res = y - cov * beta
+      val sigmaSq = (res dot res) / d
+
+      val sc = keyedRdd.sparkContext
+      val resBc = sc.broadcast(res)
+      val qtBc = sc.broadcast(q.t)
+
+      def variantPreProcess[T <: Vector[Double]](gs: T, w: Double): SkatTuple[T] = {
+        val sqrtw = math.sqrt(w.asInstanceOf[Double])
+        val wx: T = (gs * sqrtw).asInstanceOf[T]
+        val sj = resBc.value dot wx
+        SkatTuple(sj * sj, wx, qtBc.value * wx)
+      }
+
+      val skatRDD = keyedRdd
+        .map {case (k, vs) =>
+          val vArray = vs.toArray.map{ case (gs, w) => variantPreProcess(gs, w) }
+          val skatStat = if (vArray.length * n < Int.MaxValue) {
+            resultOp(vArray, sigmaSq)
+          }
+          else {
+            largeNResultOp(vArray, sigmaSq)
+          }
+          Row(k, skatStat.q, skatStat.pval, skatStat.fault)
+        }
+
+      val (skatSignature, _) = TStruct(keyName -> keyType.asInstanceOf[Type]).merge(SkatStat.schema)
```

Comment:
I also think it would simplify the interface to remove the key_name parameter in favor of using ``key``; users can always rename a column.  I was thinking to do this when I refactor the burden tests through Matrix.